src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py

```diff
@@ -0,0 +1,322 @@
+        """
+        Check conditions from previous runs of the same tool and load version
+        """
+
+        if not self._task_exists():
+            return True
+
+        latest_task = self._get_latest_task()
+        latest_task_status, job_id = latest_task['job_status'], latest_task['job_id']
+
+        if restart_on_demand:
+            # cancel the previous job if it is running or pending
+            if latest_task_status in [JobStatus.RUNNING, JobStatus.PENDING]:
+                self._cancel_job(job_id)
+
+            return True
+
+        if latest_task['source_data_dir'] != self.source_data_dir:
+            raise PreconditionError(
+                f'There is a previous run of the same tool and load version with a different source data directory.')
+
+        if latest_task_status in [JobStatus.FAILED, JobStatus.CANCELLED]:
+            print(f'The tool and load version have been run before, '
+                  f'and the most recent status is {str(latest_task_status)}.'
+                  f'Resuming progress from the previous run.')
+        elif latest_task_status in [JobStatus.COMPLETED, JobStatus.RUNNING, JobStatus.PENDING]:
+            raise PreconditionError(
+                f'There is a previous run of the same tool and load version that is {str(latest_task_status)}.')
+        else:
+            raise ValueError(f'Unexpected job status: {latest_task_status}')
+
+    def __init__(self, kbase_collection, load_ver, tool, source_data_dir, root_dir=loader_common_names.ROOT_DIR):
+        """
+        Initialize the task manager.
+
+        :param kbase_collection: KBase collection identifier name (e.g. GTDB).
+        :param load_ver: collection load version (e.g. r207.kbase.1).
+        :param tool: tool name (e.g. gtdb_tk, checkm2, etc.)
+        :param source_data_dir: source data directory.
+        :param root_dir: root directory of the collection project.
+                         Default is the ROOT_DIR defined in src/loaders/common/loader_common_names.py
+
+        """
+
+        self.kbase_collection = kbase_collection
+        self.load_ver = load_ver
+        self.tool = tool
+        self.source_data_dir = source_data_dir
+        self.root_dir = root_dir
+
+        # job directory is named as <kbase_collection>_<load_ver>_<tool>
+        self.job_dir = os.path.join(self.root_dir, TASKFARMER_JOB_DIR,
+                                    f'{self.kbase_collection}_{self.load_ver}_{self.tool}')
+
+    def submit_job(self, restart_on_demand=False):
+        """
+        Submit the job to slurm
+
+        Follow the steps in https://docs.nersc.gov/jobs/workflow/taskfarmer/#taskfarmer and generate all the necessary
+        files for the job submission (e.g. wrapper script, task list and batch script, etc.) before calling this function.
+
+        :param restart_on_demand: if True, killed any running/pending jobs and restart running the tool again.
+        """
+
+        # check if all the required files exist
```

Comment:
This method used to delete *.tfin files before it started. If it doesn't do that, won't that mess up the job status checking method, which IIUC interprets the presence of tfin files as meaning the job is successful?

Reply:
no it won't mess up job status checking. NERSC keeps track of tasks using those files. I was only testing 1 job before and thought those are not needed. But when I tested multiple jobs, those files are very helpful. Say 1 one 5 jobs failed. When we resubmit, it will once run the failed 1.

Reply:
But if you run a new job and the tfin files are still there, the task manager will think the current job is done, right, and so would allow starting another job even if `force` is false?

Reply:
force will recreate the job_dir.

Reply:
yea. in case of failed/canceled/time-out.

Reply:
Hold on, my logic is bad. What about this scenario:

* Run job, job finishes successfully, leaving `tfin` files.
* Run another job for the same workdir. This is allowed since the job is done.
* `tfin files are still there so the task manager will think the just submitted job is done
* Run another job for the same workdir. Since the task manager thinks the 2 previous jobs are both done, it'll allow the 3rd job to start even without `force`, which isn't supposed to happen

Is that right?

Reply:
`Run another job for the same workdir. This is allowed since the job is done.` When there is a previous done job, task manager will not allow another run without --force.

Reply:
Ok, you're right. In that case, should `--force` cause the task manager to error if there are tfin files, or maybe just delete them?

Reply:
force will recreate the job_dir meaning we no longer care about previous runs. tfin files from previous run are not needed any more.

Reply:
But there's no guarantee in the task manager that happens. The task manager should check that conditions are right to start the job and not rely on the calling function

Reply:
hmmm now you completely lost me. when you call `submit_job`, it does `_check_preconditions`. that function checks all conditions a job need to run.

Reply:
👍 removed all tfin files when restart_on_demand is checked.

Reply:
:+1: